cmd2/parsing.py

```diff
@@ -0,0 +1,387 @@
+#
+# -*- coding: utf-8 -*-
+"""Statement parsing classes for cmd2"""
+
+import re
+import shlex
+from typing import List, Tuple
+
+import cmd2
+
+BLANK_LINE = '\n'
+
+def _comment_replacer(match):
+    s = match.group(0)
+    if s.startswith('/'):
+        # treat the removed comment as an empty string
+        return ''
+    else:
+        return s
+
+class Statement(str):
+    """String subclass with additional attributes to store the results of parsing.
+    
+    The cmd module in the standard library passes commands around as a
+    string. To retain backwards compatibility, cmd2 does the same. However, we
+    need a place to capture the additional output of the command parsing, so we add
+    our own attributes to this subclass.
+
+    The string portion of the class contains the arguments, but not the command, nor
+    the output redirection clauses.
+    """
+    def __init__(self, object):
+        self.raw = str(object)
+        self.command = None
+        self.multilineCommand = None
+        # has to be an empty string for compatibility with standard library cmd
+        self.args = ''
+        self.terminator = None
+        self.suffix = None
+        self.pipeTo = None
+        self.output = None
+        self.outputTo = None
+    
+    @property
+    def command_and_args(self):
+        """Combine command and args with a space separating them"""
+        return '{} {}'.format('' if self.command is None else self.command, self.args).strip()
+
+class StatementParser():
+    """Parse raw text into command components.
+    
+    Shortcuts is a list of tuples with each tuple containing the shortcut and the expansion.
+    """
+    def __init__(
+            self,
+            quotes=['"', "'"],
```

Comment:
**WARNING: Strongly recommend against using mutable default arguments because they behave in an extremely non-intuitive manner nearly identical to static local variables in C/C++.**

See this reference for more info:
http://docs.python-guide.org/en/latest/writing/gotchas/

Using a tuple as the default argument value is acceptable.  If it really needs to be a mutable type like a list or dictionary, then set it to None by default and the first thing in the function check if it is None and if so set it to an empty list/dict.

This comment applies to other default argument values as well which are also mutable.

In the unlikely event that you actually want the behavior provided by mutable default arguments, then please comment to explain this.

Reply:
Yup. Fixed and pushed to `ply` branch.